### PR TITLE
Elevated vars 4994

### LINKED ISSUE #4994: Versions above 0.11.0 fail reproducably when provisioning Windows AMIs in AWS
After upgrading Packer from 0.11.0 to any newer version (incl. 1.0.0), it starts to fail reproducably when creating Windows Server 2008R2 AMIs from Amazon base AMIs in AWS. Windows Server 2012R2 AMIs are still provisioned fine. The last bits of log output are:
```
[2017-06-09T12:43:01] INFO  -- axc : ==> amazon-ebs: Restarting Machine
[2017-06-09T12:43:01] INFO  -- axc : ==> amazon-ebs: Waiting for machine to restart...
[2017-06-09T12:43:01] INFO  -- axc :     amazon-ebs: A system shutdown is in progress.(1115)
[2017-06-09T12:43:53] INFO  -- axc :     amazon-ebs: WIN-2E19G4EGGMK restarted.
[2017-06-09T12:43:53] INFO  -- axc : ==> amazon-ebs: Machine successfully restarted, moving on
[2017-06-09T12:43:53] INFO  -- axc : ==> amazon-ebs: Provisioning with Powershell...
[2017-06-09T12:43:53] INFO  -- axc : ==> amazon-ebs: Provisioning with shell script: /path/to/install_puppet.ps1
[2017-06-09T12:44:24] INFO  -- axc : ==> amazon-ebs: Terminating the source AWS instance...
[2017-06-09T12:44:55] INFO  -- axc : ==> amazon-ebs: Cleaning up any extra volumes...
[2017-06-09T12:44:55] INFO  -- axc : ==> amazon-ebs: No volumes to clean up, skipping
[2017-06-09T12:44:55] INFO  -- axc : ==> amazon-ebs: Deleting temporary keypair...
[2017-06-09T12:44:55] INFO  -- axc : Build 'amazon-ebs' errored: Script exited with non-zero exit status: 2147942487. Allowed exit codes are: [0]
[2017-06-09T12:44:55] INFO  -- axc :
[2017-06-09T12:44:55] INFO  -- axc : ==> Some builds didn't complete successfully and had errors:
[2017-06-09T12:44:55] INFO  -- axc : --> amazon-ebs: Script exited with non-zero exit status: 2147942487. Allowed exit codes are: [0]
[2017-06-09T12:44:55] INFO  -- axc :
[2017-06-09T12:44:55] INFO  -- axc : ==> Builds finished but no artifacts were created.
```

- Versions: 0.12.x, 1.0.0
- Ubuntu Linux
- https://gist.github.com/dhs-rec/a5cf71c78ba7f8a832ab0fcb8d49483e

### PR DESCRIPTION
**Pull request by SwampDragons**
Source environment variables from a file, instead of writing to command line.  
Simplify upload of elevated powershell wrapper.
Finish implementation of elevated_execute_command option in template.
TODO: 
- [x] Update docs.
- [x] Fix tests

Closes #4994

### CODE REVIEW

File: provisioner/powershell/provisioner.go [hunk 1]
```diff
@@ -458,27 +469,12 @@ func (p *Provisioner) generateElevatedRunner(command string) (uploadedPath strin
...
-	}
-	tmpFile.Close()
-	f, err := os.Open(tmpFile.Name())
-	if err != nil {
-		return "", fmt.Errorf("Error opening temporary elevated powershell script: %s", err)
-	}
-	defer f.Close()
-
+	wrapperBytes := buffer.Bytes()
+	wrapperReader := bytes.NewReader(wrapperBytes)
```
Comment (mwhooker) [inline provisioner/powershell/provisioner.go]: I think `buffer.Bytes()` already implements the reader interface, so I don't think this is needed

File: provisioner/powershell/provisioner.go [hunk 2]
```diff
@@ -458,27 +469,12 @@ func (p *Provisioner) generateElevatedRunner(command string) (uploadedPath strin
...
 	uuid := uuid.TimeOrderedUUID()
 	path := fmt.Sprintf(`${env:TEMP}\packer-elevated-shell-%s.ps1`, uuid)
-	log.Printf("Uploading elevated shell wrapper for command [%s] to [%s] from [%s]", command, path, tmpFile.Name())
-	err = p.communicator.Upload(path, f, nil)
+	log.Printf("Uploading elevated shell wrapper for command [%s] to [%s]", command, path)
+	err = p.communicator.Upload(path, wrapperReader, nil)
```
Comment (mwhooker) [inline provisioner/powershell/provisioner.go]: I hadn't even thought to just use the buffer directly, nice!

**Comment by mwhooker** [comment]
some tests failures but I think this definitely the right idea

File: provisioner/powershell/provisioner.go (hunk 1 above)
Comment (SwampDragons) [reply provisioner/powershell/provisioner.go]: Good thought, but it appears not: `cannot use buffer (type bytes.Buffer) as type io.Reader in argument to p.communicator.Upload: bytes.Buffer does not implement io.Reader (Read method has pointer receiver)`

Comment (mwhooker) [reply provisioner/powershell/provisioner.go]: I would double check that

https://play.golang.org/p/5Tx2KPcucj

**Pull request merged**